### ISSUE TITLE
Sealing with Blake2s (was: Hybrid Merkle Trees)

### ISSUE THREAD

**Issue by porcuquine** [filecoin-project/rust-fil-proofs#531]
### Description
From #77:

---
**UPDATE: note 'pending calculations' below. It may actually be the case that the optimal balance is to use ALL blake2s — in which case the cost in development time and residual complexity of the hybrid solution would be wasted.**

**Based on calculations so far, I tentatively think the next step should be to get blake2s working in circuits (~generically~). This is a more manageable next step that would have to be accomplished anyway.**

**Since this might be the right answer, we could perhaps do this not by being generic but by just swapping blake2s in for pedersen hashing. As I recall, this would sidestep the complexity and be much simpler. That would let us perform tests to validate assumptions as quickly as posible.**

---
**UPDATE 2: calculations still need to be checked more carefully, but tentative conclusions are clear.**

![image](https://user-images.githubusercontent.com/1746729/54076988-26b56e80-4267-11e9-97c4-fe70f39190a5.png)

**Fuller analysis is forthcoming.**

Newer graphs suggest Blake2s advantage kicks in around 1TB, in which case we really need to run some pure blake tests (including circuits).

![image](https://user-images.githubusercontent.com/1746729/54100276-f232d700-437a-11e9-85cd-2bea33c63f6c.png)

Obviously one issue is that for blake2s to be the better solution, we need (with current numbers) an enormous sector (> 2 TiB — with which our current rules means ~4TiB~ UPDATE: although more favorable to blake2s, new numbers indicate even larger sector sizes to justify seal times). This is too big, so it's not yet clear that blake2s is the answer — but it's now the best guess. LATER UPDATE: More calculation suggest that if we can get groth proving time down to projected time, we can make this work with smaller sectors — maybe 256GiB, which may be viable. It's currently much higher, but there's a lot of noise with parameter generation, loading, perhaps something with how we parallelize. Worst case, at that size, we can break into many more circuits and run on much smaller machines.

One thing, however, that these graphs do make clear (if accurate) is that the hybrid solution is not actually better in any circumstances. Before the intersection point at which blake2s and pedersen perform the same, pedersen is better; after, blake is.

### UPDATE conclusion:
**Therefore, we can (for now) conclude that the hybrid solution provides no advantages. We should instead focus on being able to swap blake2s in for pedersen in circuits so we can get some real performance and circuit data to improve our projections as we solve for a viable configuration.**

---
>  Pending calculations verifying a projected solution to the performance equation, implement a hybrid hashing solution for merkle trees. The idea is to allow a tunable public parameter (a tree depth) specifying the portion of the tree which should use blake2s instead of pedersen hashes. This will allow us to trade proof size for speed. This is exactly the lever we need to take advantage of our surplus proof-size budget and apply it to cpu-time performance.

Here's an approach I think can work. If you have a better idea or aren't sure about this, let's discuss before beginning:

  - Implement a new hybrid merkle tree type, say `HybridMerkleTree`, parameterized by two `Hash` types.
  - For this description, call the types `RootHash` and `LeafHash`.
  - Include a field called `root_tree_height` corresponding to the number of rows which should be contained in the `RootHash` tree.
  - The new type should implement all the methods of `merkle_light::MerkleTree` (or at least all the ones we use).
  - When constructing a `HybridMerkleTree`, create `N` `MerkleTree<LeafHash,…>` trees.
  - The root of each of these leaf trees will be one of the leaves of the single `HybridMerkleTree<RootHash,…>` tree.
  - The leaves of all the leaf trees will be the leaves of the hybrid tree.
  - The root of the root tree will be the root of the hybrid tree.

The rest should follow naturally from this description. Example:
 - Say you have a `HybridMerkleTree` of height 10.
 - `root_tree_height = 3`: the root tree will have three rows, and the leaf row will have 4 leaves.
 - Therefore, there will be 4 leaf trees.
 - Each leaf tree will be of height 8.
 - Root tree height + leaf tree height = hybrid tree height - 1.
 - The '- 1` is for the overlap between (unhashed) root-tree leaves and leaf-tree roots, which are the same value.

Once `HybridMerkleTree` is implemented:
 - Refactor the plumbing so `ZigZagDrgPoRep` allows a hybrid tree.  (yes, adding the extra type parameter will be annoying).
 - Add `root_tree_height` to `layered_drgporep::PublicParams` — default 0.
 - This implies that to mimic the behavior of our current `ZigZagDrgPorep`, the `LeafHash` should be `DefaultTreeHasher` (`PedersenHasher`).
 - For consistency, maybe we should also create a type alias for `DefaultRootTreeHasher = Blake2sHasher`.

Last but not even vaguely least, extend circuits to make this work. This probably belongs in a follow-up. Feel free to skip and create a new issue for the circuit component once the vanilla part is done.

Optional for now, but will be needed immediately in a follow-up if not implemented now:
  - Extend zigzag example to allow specification of a root hash and root tree height, using them in the obvious way.

I think that's about it. A lot, I know, but I think it's mostly mechanical if you understand the design.

### Acceptance criteria

  - Everything above is implemented as described (modulo naming, but let's discuss proposed changes).

### Risks + pitfalls
  - There will be a lot of mechanical refactoring in adding the type parameters.
  - I probably failed to anticipate something particularly hairy.
  - I *think* if you stay close to the design, things should shake out — so if it feels like a lot of deviation and innovation beyond dodging the type system is needed, let's discuss.

### Where to begin
 
I'm tempted to change it up and say, 'Sew, a needle pulling thread,' but we should probably keep it simple and start at the beginning, as described above.

**Comment by nicola** [comment]
I am not sure if I follow correctly, does the calculation say it's not worth to implement this, but we are still implementing it anyway to test?

**Comment by porcuquine** [comment]
No, we should not implement the hybrid. We should swap in blake2s as the default merkle-tree hasher (ideally in such a way that it's easy to switch back if we need to). I will create a new issue for that, but for now the justification lives here.

**Comment by dignifiedquire** [comment]
We have the functionality done for blake2s sealing
